proxy/envoy/discovery.go

```diff
@@ -15,24 +15,42 @@
 package envoy
 
 import (
+	"encoding/json"
 	"fmt"
 	"net/http"
 	"net/http/pprof"
 	"strconv"
+	"sync"
 
 	restful "github.com/emicklei/go-restful"
 	"github.com/golang/glog"
+	"github.com/golang/protobuf/proto"
 
 	proxyconfig "istio.io/api/proxy/v1/config"
 	"istio.io/manager/model"
 )
 
 // DiscoveryService publishes services, clusters, and routes for all proxies
 type DiscoveryService struct {
-	services model.ServiceDiscovery
-	config   *model.IstioRegistry
-	mesh     *proxyconfig.ProxyMeshConfig
-	server   *http.Server
+	services   model.ServiceDiscovery
+	controller model.Controller
+	config     *model.IstioRegistry
+	mesh       *proxyconfig.ProxyMeshConfig
+	server     *http.Server
+
+	disableCache bool
+	cacheMu      sync.Mutex
```

Comment:
I considered RWMutex here but we'd need to grab the write lock to update the stats anyway. We could add two locks -- one for stats and another for the cache -- but then we'd be grabbing a write *and* read lock per request. Instead, I refactored things a bit so the lock is held for a shorter duration in the hot path. The slow path will take the long twice and there is a small possibility we could regenerate the non-cached response twice, but that shouldn't cause any functional problems, only a bit of a perf bit. This may or may not be the right trade off for hot vs. slow path, but we can continue to profile this area if necessary as we start to track overall performance of the system.